qiskit/primitives/base_estimator.py

```diff
@@ -0,0 +1,199 @@
+
+The output is the expectation value or a list of expectation values.
+
+.. math::
+
+    \Braket{\psi_i(\theta_k)|H_j|\psi_i(\theta_k)}
+
+
+The estimator object is expected to be initialized with "with" statement
+and the objects are called with parameter values and run options
+(e.g., ``shots`` or number of shots).
+
+Here is an example of how estimator is used.
+
+.. code-block:: python
+
+    from qiskit.circuit.library import RealAmplitudes
+    from qiskit.quantum_info import SparsePauliOp
+
+    psi1 = RealAmplitudes(num_qubits=2, reps=2)
+    psi2 = RealAmplitudes(num_qubits=2, reps=3)
+
+    params1 = psi1.parameters
+    params2 = psi2.parameters
+
+    H1 = SparsePauliOp.from_list([("II", 1), ("IZ", 2), ("XI", 3)])
+    H2 = SparsePauliOp.from_list([("IZ", 1)])
+    H3 = SparsePauliOp.from_list([("ZI", 1), ("ZZ", 1)])
+
+    with Estimator([psi1, psi2], [H1, H2, H3], list(params1) + list(params2)) as e:
+        # first 6 values correspond to the parameters of psi1
+        # last 8 values correspond to the parameters of psi2
+        theta1 = [0, 1, 1, 2, 3, 5] + [0] * 8
+        theta2 = [0] * 6 + [0, 1, 1, 2, 3, 5, 8, 13]
+        theta3 = [1, 2, 3, 4, 5, 6] + [0] * 8
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)> ]
+        psi1_H1_result = e(theta1, shots=1024, grouping=[(0, 0)])
+        print(psi1_H1_result)
+
+        # calculate [ <psi1(theta1)|H2|psi1(theta1)>, <psi1(theta1)|H3|psi1(theta1)> ]
+        psi1_H23_result = e(theta1, shots=1024, grouping=[(0, 1), (0, 2)])
+        print(psi1_H23_result)
+
+        # calculate [ <psi2(theta2)|H2|psi2(theta2)> ]
+        psi2_H2_result = e(theta2, shots=1024, grouping=[(1, 1)])
+        print(psi2_H2_result)
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)>, <psi1(theta3)|H1|psi1(theta3)> ]
+        psi1_H1_result2 = e([theta1, theta3], shots=1024, grouping=[(0, 0)])
+        print(psi1_H1_result2)
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)>, <psi2(theta2)|H2|psi2(theta2)>, <psi1(theta3)|H3|psi1(theta3)> ]
+        psi12_H23_result = e([theta1, theta2, theta3], shots=1024, grouping=[(0, 0), (1, 1), (0, 2)])
+        print(psi12_H23_result)
+
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Optional, Union
+
+from qiskit.circuit import QuantumCircuit, Parameter
+from qiskit.circuit.parametertable import ParameterView
+from qiskit.quantum_info import SparsePauliOp
+
+from .estimator_result import EstimatorResult
+
+
+@dataclass(frozen=True)
+class Group:
+    """The dataclass represents indices of circuit and observable."""
+
+    circuit_index: int
+    observable_index: int
+
+
+class BaseEstimator(ABC):
+    """Estimator base class.
+
+    Base class for Estimator that estimates expectation values of quantum circuits and observables.
+    """
+
+    def __init__(
+        self,
+        circuits: list[QuantumCircuit],
+        observables: list[SparsePauliOp],
```

Comment:
PauliSumOp is not allowed here?

Reply:
I can support it, but I think it should be implementation dependent, not ABC.

Supporting PauliSumOp in the base class makes it difficult because it forces the subclass to also support PauliSumOp.